### ISSUE TITLE
No consume body when no content

### ISSUE THREAD

**Issue by duylong** [logstash-plugins/logstash-output-http#6]
Hi,

It seems when I get a 204 code (No Content), Logstash waits indefinitely. The first request sent is received.

No problem in ignoring read_body function in this case:

```
      # Consume body to let this connection be reused
      #rbody = ""
      #response.read_body { |c| rbody << c }
```
Cordialy,

**Comment by duylong** [comment]
In ignoring read_body, I have a OutOfMemoryError in Logstash.

A response example:
```
HTTP/1.1 204 No Content
request-id: 07124baf-0f44-11e5-801e-000000000000
x-app-version: FIXME
date: Wed, 10 Jun 2015 07:40:55 GMT
transfer-encoding: chunked
```

**Comment by sguiheux** [comment]
Hello, 

Same error here. It seems that logstash wait for 1mn by line to send  and log this:
Unhandled exception {
	:request=><FTW::Request(@13792) 
	@headers=FTW::HTTP::Headers <{
		"host"=>"XXXXX", 
		"connection"=>"keep-alive", 
		"authorization"=>"Basic XXXXX", 
		"content-type"=>"application/json", 
		"content-length"=>91}> 
	@method="POST" 
	@body="{\"metric\":\"test.jvm\",\"timestamp\":1439972594,\"value\":\"0.0114187\", \"tags\":{\"host\":\"desksgu\"}}" 
	@logger=#<Cabin::Channel:0x9fcd2 
	@metrics=#<Cabin::Metrics:0x7a5fa911 
	@metrics_lock=#<Mutex:0x2620b569>, 
	@metrics={}, 
	@channel=#<Cabin::Channel:0x9fcd2 ...>>, 
	@subscriber_lock=#<Mutex:0x66f081f9>, 
	@level=:info, @subscribers={}, 
	@data={}> 
	@request_uri="/api/put" 
	@version=1.1 
	@port=80 
	@protocol="http" >, 
	
	:response=>#<FTW::Response:0x5e5b1c6b 
	@reason="No Content", 
	@version=1.1, 
	@headers=FTW::HTTP::Headers <{
		"access-control-allow-origin"=>"*", 
		"date"=>"Wed, 19 Aug 2015 08:27:10 GMT", 
		"transfer-encoding"=>"chunked"}>, 
	@body=<FTW::Connection(@13794) 
	@destinations=["XXXXX"] 
	@connected=false 
	@remote_address="XXXXX" 
	@secure=false >, 
	@logger=#<Cabin::Channel:0x9fcd2 
	@metrics=#<Cabin::Metrics:0x7a5fa911 
	@metrics_lock=#<Mutex:0x2620b569>, 
	@metrics={}, 
	@channel=#<Cabin::Channel:0x9fcd2 ...>>, 
	@subscriber_lock=#<Mutex:0x66f081f9>, 
	@level=:info, 
	@subscribers={}, 
	@data={}>, 
	@status=204>, 
	:exception=>#<EOFError: End of file reached>, :stacktrace=>[
		"org/jruby/RubyIO.java:3007:in `sysread'", 
		"/home/sguiheux/Soft/logstash-1.5.3/vendor/bundle/jruby/1.9/gems/ftw-0.0.44/lib/ftw/connection.rb:247:in `read'", 
		"/home/sguiheux/Soft/logstash-1.5.3/vendor/bundle/jruby/1.9/gems/ftw-0.0.44/lib/ftw/protocol.rb:196:in `read_http_body_chunked'", 
		"/home/sguiheux/Soft/logstash-1.5.3/vendor/bundle/jruby/1.9/gems/ftw-0.0.44/lib/ftw/protocol.rb:127:in `read_http_body'", 
		"/home/sguiheux/Soft/logstash-1.5.3/vendor/bundle/jruby/1.9/gems/ftw-0.0.44/lib/ftw/protocol.rb:149:in `read_body'", 
		"/home/sguiheux/Soft/logstash-1.5.3/vendor/bundle/jruby/1.9/gems/logstash-output-http-1.0.0/lib/logstash/outputs/http.rb:130:in `receive'", 
		"/home/sguiheux/Soft/logstash-1.5.3/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.3-java/lib/logstash/outputs/base.rb:88:in `handle'", 
		"(eval):106:in `output_func'", 
		"/home/sguiheux/Soft/logstash-1.5.3/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.3-java/lib/logstash/pipeline.rb:244:in `outputworker'", 
		"/home/sguiheux/Soft/logstash-1.5.3/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.3-java/lib/logstash/pipeline.rb:166:in `start_outputs'"
	], 
	:level=>:warn, :file=>"logstash/outputs/http.rb", :line=>"133", :method=>"receive"}

**Comment by growse** [comment]
I'm seeing a similar issue when using the http plugin to output to influxdb. Influx outputs a 204 response with no content, which causes logstash to wait.

Looks like the same exception as above, read_body causes an `EOFError` to be thrown.

**Comment by andrewvc** [comment]
Looks like we need to fix this error here: https://github.com/logstash-plugins/logstash-output-http/blob/master/lib/logstash/outputs/http.rb#L130

**Comment by andrewvc** [comment]
@jordansissel is this code that consumes the body strictly necessary? https://github.com/logstash-plugins/logstash-output-http/blob/master/lib/logstash/outputs/http.rb#L128

Seems to be the culprit here. Ideally we would just never perform the read

**Comment by growse** [comment]
FWIW, I've hacked around it for the moment by doing this:

    if response.status != 204
        response.read_body { |c| rbody << c }
    end

This is probably a bad solution though, because there could be other status types where the Content-Length==0. I couldn't see an obvious place to find Content-Length from FTW though.

**Comment by andrewvc** [comment]
This should be fixed with the latest release (1.1.0).

If this is still an issue please reopen this ticket!
